### PR TITLE
OCPBUGS-11304: daemon: event only on actual OS updates

### PR DESCRIPTION
**Pull request by yuqi-zhang**
Our OSUpdateStarted/OSUpdateStaged were being thrown when we entered the function, which I think is a bit confusing if no OS updates actually were to happen.

This aims to fix the CI issue where the MCO events were being mis-interpreted as the update stage taking too long, although we do potentially open ourselves to the issue of a non-osupdate-upgrade (i.e. the payloads have the same OSImage), but that should be rare, and I am not sure if no events were emitted, if we'd fail or not. I guess we can find out if we ever hit it.

### CODE REVIEW

**Comment by yuqi-zhang** [comment]
/jira refresh

you'd think an unreleased version of 4.14 would equate to 4.14.0

File: pkg/daemon/update.go [hunk 1]
```diff
@@ -375,40 +375,39 @@ func (dn *CoreOSDaemon) applyOSChanges(mcDiff machineConfigDiff, oldConfig, newC
...
+				nodeName = dn.node.ObjectMeta.GetName()
+				nodeObjRef = corev1.ObjectReference{
+					Kind: "Node",
+					Name: dn.node.GetName(),
+					UID:  dn.node.GetUID(),
+				}
+			}
+			// We send out the event OSUpdateStaged synchronously to ensure it is recorded.
+			// Remove this when we can ensure all events are sent before exiting.
+			t := metav1.NewTime(time.Now())
```
Comment (Prashanth684) [inline pkg/daemon/update.go:391]: the way the two events are created are still different though? is that ok?

**Comment by sergiordlr** [comment]
Verified using IPI on GCP

1. Remove all events in "default" namespace to avoid any possible mistake
```
$ oc delete events -n default --all
```
2. Create a MC that will not update the OS (no osimage, no kernel args, no kernel type and no extensions)

We will use a passwd MC:

```	
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name: test-machine-config-passwdhash
spec:
  config:
    ignition:
      version: 3.2.0
    passwd:
      users:
      - name: core
        passwordHash: $6$uim4LuKWqiko1l5K$QJUwg.4lAyU4egsM7FNaNlSbuI6JfQCRufb99QuF082BpbqFoHP3WsWdZ5jCypS0veXWN1HDqO.bxUpE9aWYI1
```

Wait until the configuration is applied in all nodes


3. Check that no OSUpdateStarted and no OSUpdateStaged events were triggered
```
$ oc get events -n default --sort-by metadata.creationTimestamp --field-selector reason=OSUpdateStarted
No resources found in default namespace.

$ oc get events -n default --sort-by metadata.creationTimestamp --field-selector reason=OSUpdateStaged
No resources found in default namespace.
```
Check the events for a given node (no OSUpdate):
```
$ oc get events -n default --sort-by metadata.creationTimestamp --field-selector involvedObject.name=sregidor-osver2-fhnz6-worker-a-bj7vq.c.openshift-qe.internal
LAST SEEN   TYPE     REASON                      OBJECT                                                              MESSAGE
3m2s        Normal   ConfigDriftMonitorStopped   node/sregidor-osver2-fhnz6-worker-a-bj7vq.c.openshift-qe.internal   Config Drift Monitor stopped
3m          Normal   SkipReboot                  node/sregidor-osver2-fhnz6-worker-a-bj7vq.c.openshift-qe.internal   Config changes do not require reboot.
3m          Normal   PendingConfig               node/sregidor-osver2-fhnz6-worker-a-bj7vq.c.openshift-qe.internal   Written pending config rendered-worker-cd2a60dd1fcdcd64f365d0be509c6cb3
2m50s       Normal   Uncordon                    node/sregidor-osver2-fhnz6-worker-a-bj7vq.c.openshift-qe.internal   Update completed for config rendered-worker-cd2a60dd1fcdcd64f365d0be509c6cb3 and node has been uncordoned
2m50s       Normal   NodeDone                    node/sregidor-osver2-fhnz6-worker-a-bj7vq.c.openshift-qe.internal   Setting node sregidor-osver2-fhnz6-worker-a-bj7vq.c.openshift-qe.internal, currentConfig rendered-worker-cd2a60dd1fcdcd64f365d0be509c6cb3 to Done
2m50s       Normal   ConfigDriftMonitorStarted   node/sregidor-osver2-fhnz6-worker-a-bj7vq.c.openshift-qe.internal   Config Drift Monitor started, watching against rendered-worker-cd2a60dd1fcdcd64f365d0be509c6cb3
```



4. Create a MC that will update the OS (we will use a kernel args update)

```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name: test-kernel-arguments-32
spec:
  config:
    ignition:
      version: 3.2.0
  kernelArguments:
    - loglevel=7
```

Wait for all the nodes to be updated

5. Check that OSUpdate events were triggered for worker nodes

```
$  oc get events -n default --sort-by metadata.creationTimestamp --field-selector reason=OSUpdateStarted
LAST SEEN   TYPE     REASON            OBJECT                                                              MESSAGE
10m         Normal   OSUpdateStarted   node/sregidor-osver2-fhnz6-worker-a-bj7vq.c.openshift-qe.internal   Changing kernel arguments
7m17s       Normal   OSUpdateStarted   node/sregidor-osver2-fhnz6-worker-b-dx86n.c.openshift-qe.internal   Changing kernel arguments
4m25s       Normal   OSUpdateStarted   node/sregidor-osver2-fhnz6-worker-c-c55zk.c.openshift-qe.internal   Changing kernel arguments

$  oc get events -n default --sort-by metadata.creationTimestamp --field-selector reason=OSUpdateStaged
LAST SEEN   TYPE     REASON           OBJECT                                                              MESSAGE
10m         Normal   OSUpdateStaged   node/sregidor-osver2-fhnz6-worker-a-bj7vq.c.openshift-qe.internal   Changes to OS staged
7m46s       Normal   OSUpdateStaged   node/sregidor-osver2-fhnz6-worker-b-dx86n.c.openshift-qe.internal   Changes to OS staged
4m54s       Normal   OSUpdateStaged   node/sregidor-osver2-fhnz6-worker-c-c55zk.c.openshift-qe.internal   Changes to OS staged
```
Verify the events generated for a single node:
```
$ oc get events -n default --sort-by metadata.creationTimestamp --field-selector involvedObject.name=sregidor-osver2-fhnz6-worker-a-bj7vq.c.openshift-qe.internal |grep -E "OSUpdateStarted|OSUpdateStaged"
11m         Normal    OSUpdateStarted             node/sregidor-osver2-fhnz6-worker-a-bj7vq.c.openshift-qe.internal   Changing kernel arguments
11m         Normal    OSUpdateStaged              node/sregidor-osver2-fhnz6-worker-a-bj7vq.c.openshift-qe.internal   Changes to OS staged
```

6. Check events after upgrade from 4.13 
```
$ oc get clusterversion
NAME      VERSION                              AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.13.0-0.nightly-2023-05-04-090524   True        False         48m     Cluster version is 4.13.0-0.nightly-2023-05-04-090524
```
Remove all envents from default namespace before upgrading

Execute the upgrade
```
$ oc get clusterversion -o yaml
...
    history:
    - acceptedRisks: |-
        Target release version="" image="registry.build05.ci.openshift.org/ci-ln-tnrrm92/release:latest" cannot be verified, but continuing anyway because the update was forced: release images that are not accessed via digest cannot be verified
        Precondition "ClusterVersionRecommendedUpdate" failed because of "UnknownUpdate": RetrievedUpdates=False (VersionNotFound), so the recommended status of updating from 4.13.0-0.nightly-2023-05-04-090524 to 4.13.0-0.ci.test-2023-05-05-074737-ci-ln-tnrrm92-latest is unknown.
      completionTime: "2023-05-05T12:03:24Z"
      image: registry.build05.ci.openshift.org/ci-ln-tnrrm92/release:latest
      startedTime: "2023-05-05T11:04:12Z"
      state: Completed
      verified: false
      version: 4.13.0-0.ci.test-2023-05-05-074737-ci-ln-tnrrm92-latest
    - completionTime: "2023-05-05T10:13:05Z"
      image: registry.ci.openshift.org/ocp/release@sha256:8a6390a385ab8a138d33d25afa48ac7351980c852f748b689779ba649dcee937
      startedTime: "2023-05-05T09:49:49Z"
      state: Completed
      verified: false
      version: 4.13.0-0.nightly-2023-05-04-090524
```

Check that OSUpdate events were properly triggered:

```
$ oc get events -n default --sort-by metadata.creationTimestamp --field-selector involvedObject.name=sregidor-orig3-jknxj-worker-a-57qtq.c.openshift-qe.internal | grep OSU
9m21s       Normal    OSUpdateStarted             node/sregidor-orig3-jknxj-worker-a-57qtq.c.openshift-qe.internal   Upgrading OS
9m6s        Normal    OSUpgradeApplied            node/sregidor-orig3-jknxj-worker-a-57qtq.c.openshift-qe.internal   OS upgrade applied; new MachineConfig (rendered-worker-13935fb4a16c5d7ff426dd95d024638b) has new OS image (registry.build05.ci.openshift.org/ci-ln-tnrrm92/stable@sha256:ecfed6de7b25ae1007fe35af942c50310f5c59f1ca613fe2fee9594a4de5f720)
9m6s        Normal    OSUpdateStaged              node/sregidor-orig3-jknxj-worker-a-57qtq.c.openshift-qe.internal   Changes to OS staged
```

/label qe-approved

NOTE: We need to take 2 things into account:
- As Jerry mentioned in the jira issue. Upgrades to a version with the same osImage will not trigger any OSUpdate event. If there is any test checking this event on upgrades, it may fail.
- We still can generate OSUpdateStart and OSUpdateStaged events with the same timestamp (in the same second), and those events that happened in the same second can be reported in any order, so the test cases need to be aware of that.

File: pkg/daemon/update.go (hunk 1 above)
Comment (yuqi-zhang) [reply pkg/daemon/update.go:391]: We'll try to make that consistent later. For now, this should make it such that on each upgrade, the event only shows up once each

**Review by sinnykumari — COMMENTED** [review]
We do emit later on either OSUpgradeApplied or OSUpgradeSkipped event to indicate whether the OSUpdate involved upgrade or not. I am wondering if  this not sufficient to avoid the confusion?

**Comment by yuqi-zhang** [comment]
> We do emit later on either [OSUpgradeApplied or OSUpgradeSkipped event ](https://github.com/openshift/machine-config-operator/blob/master/pkg/daemon/update.go#L2155)to indicate whether the OSUpdate involved upgrade or not. I am wondering if this not sufficient to avoid the confusion?

Good question, I think we would need to beef up those events as well. Today, there are the following scenarios:

1. no os update at all -> we get OSUpdateStarted and OSUpdateStaged
2. no image update, but kargs/etc. update -> we get OSUpdateStarted, OSUpgradeSkipped, OSUpdateStaged (I believe)
3. yes os update -> we get OSUpdateStarted, OSUpgradeApplied and OSUpdateStaged

So there are multiple scenarios gated on many different condition checks. I can create a card to improve that as well, but right now we are hitting scenario 1

**Comment by yuqi-zhang** [comment]
This affects only 4.13/4.14 so

/cherry-pick release-4.13

**Pull request merged**